GTBitsOfGood/Healing4Heroes
# Migrate Video Storage to AWS S3

## What is this?

This PR migrates video storage to AWS S3. Video upload and download are done client-side using presigned URLs generated by the server.

- `h4h-app` is the IAM user generating the presigned URLs. 
   - The presigned URLs inherit all IAM permissions of the user creating them so we have to be prudent about granting permissions to `h4h-app`. 
   - Currently, `h4h-app` has no login and can only perform the `GetObject`, `PutObject`, and `AbortMultipartUpload` functions on the `healing4heroes` bucket.
- Presigned URLs expire after 15 minutes.
- Files are broken down into 8MB chunks and uploaded via multipart upload to optimize processing time. 
- All video files will be stored in Glacier Instant Retrieval storage class, which is done by presigning the URLs with the header `X-Amz-Storage-Class=GLACIER_IR`. 
   - Without this header, all file uploads will default to be in the Standard storage class. 
   - I have created a fallback lifecycle policy in case the header is not set, which will move the file into Glacier Instant Retrieval after 1 day.
- Minimal error handling exists at this point. 
   - For example, there is no mechanism to handle errors with an upload and prompt the user to upload again. 
   - We might also want to implement a mechanism to abort the uploaded parts of an incomplete upload to prevent extra charges. I have mitigated this by creating a lifecycle policy in which all incomplete uploads will be deleted after 1 day.

I followed this [tutorial](https://www.altostra.com/blog/multipart-uploads-with-s3-presigned-url) to set up multipart uploading. Any questions are welcomed.

Further documentation:
- [AWS Documentation on Multipart Upload](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)
- [AWS Documentation on Presigned URL](https://docs.aws.amazon.com/AmazonS3/latest/userguide/PresignedUrlUploadObject.html)
- [AWS Javascript SDK for S3 client](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/globals.html)
- [AWS Javascript SDK for Generating Presigned URLs](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/modules/_aws_sdk_s3_request_presigner.html)
- [AWS S3 Pricing per Storage Class](https://aws.amazon.com/s3/pricing/)

## Setup



## Steps to Test


- Use the `StorageExampleScreen` to upload a file. Visit the `healing4heroes` bucket to check if the file was successfully uploaded.
- Use `/api/admin/file/download` endpoint to test download a file. The body is the key of the video file, which is everything that follows the bucket name in the file's ARN. For example,
```
{
    "key": "TrainingLogVideos/8b098274-f38e-4201-bb0d-9ab51b08266a.mp4"
}
```

## Associated Issues

Resolves #50

## Additional Instructions
